packages/core/src/event_delegation_utils.ts

```diff
@@ -0,0 +1,113 @@
+
+declare global {
+  var ngContracts: {[key: string]: EarlyJsactionDataContainer};
+  interface Element {
+    __jsaction_fns: Map<string, Function[]> | undefined;
+  }
+}
+
+export const CONTRACT_PROPERTY = 'ngContracts';
+
+export function invokeRegisteredListeners(event: Event) {
+  const handlerFns = (event.currentTarget as Element)?.__jsaction_fns?.get(event.type);
+  if (!handlerFns) {
+    return;
+  }
+  for (const handler of handlerFns) {
+    handler(event);
+  }
+}
+
+export function setJSActionAttribute(nativeElement: Element, eventTypes: string[]) {
+  if (!eventTypes.length) {
+    return;
+  }
+  const parts = eventTypes.reduce((prev, curr) => prev + curr + ':;', '');
+  const existingAttr = nativeElement.getAttribute(Attributes.JSACTION);
+  //  This is required to be a module accessor to appease security tests on setAttribute.
+  nativeElement.setAttribute(Attributes.JSACTION, `${existingAttr ?? ''}${parts}`);
+}
+
+export const sharedStashFunction = (rEl: RElement, eventType: string, listenerFn: () => void) => {
+  const el = rEl as unknown as Element;
+  const eventListenerMap = el.__jsaction_fns ?? new Map();
+  const eventListeners = eventListenerMap.get(eventType) ?? [];
+  eventListeners.push(listenerFn);
+  eventListenerMap.set(eventType, eventListeners);
+  el.__jsaction_fns = eventListenerMap;
+};
+
+export const removeListeners = (el: Element) => {
+  el.removeAttribute(Attributes.JSACTION);
+  el.__jsaction_fns = undefined;
+};
+
+// TODO: Upstream this back into event-dispatch.
+function getJsactionData(container: EarlyJsactionDataContainer) {
+  return container._ejsa;
+}
+
+@Injectable({providedIn: 'root'})
+export class GlobalEventDelegation {
+  eventContract!: EventContract;
+  addEvent(el: Element, eventName: string) {
+    if (this.eventContract) {
+      this.eventContract.addEvent(eventName);
+      setJSActionAttribute(el, [eventName]);
+      return true;
+    }
+    return false;
+  }
+}
+
+export const initEventReplay = (eventDelegation: GlobalEventDelegation, injector: Injector) => {
+  const appId = injector.get(APP_ID);
+  // This is set in packages/platform-server/src/utils.ts
+  // Note: globalThis[CONTRACT_PROPERTY] may be undefined in case Event Replay feature
+  // is enabled, but there are no events configured in an application.
+  const container = globalThis[CONTRACT_PROPERTY]?.[appId];
+  const earlyJsactionData = getJsactionData(container)!;
+  eventDelegation.eventContract = new EventContract(
+    new EventContractContainer(earlyJsactionData.c),
+  );
+  for (const et of earlyJsactionData.et) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  for (const et of earlyJsactionData.etc) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  eventDelegation.eventContract.replayEarlyEvents(container);
+  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
+  registerDispatcher(eventDelegation.eventContract, dispatcher);
```

Comment:
nit: for better compression (since we don't rename properties), we can create a const and use it like this:
```suggestion
  const eventContract = new EventContract(
    new EventContractContainer(earlyJsactionData.c),
  );
  for (const et of earlyJsactionData.et) {
    eventContract.addEvent(et);
  }
  for (const et of earlyJsactionData.etc) {
    eventContract.addEvent(et);
  }
  eventContract.replayEarlyEvents(container);
  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
  registerDispatcher(eventContract, dispatcher);
  eventDelegation.eventContract = eventContract;
```